fem/hyperbolic_conservation_laws.hpp

```diff
@@ -0,0 +1,498 @@
+    */
+   inline void resetMaxCharSpeed()
+   {
+      max_char_speed = 0.0;
+   }
+
+   inline double getMaxCharSpeed()
+   {
+      return max_char_speed;
+   }
+
+   /**
+    * @brief implement (F(u), grad v) with abstract F computed by ComputeFlux
+    *
+    * @param[in] el local finite element
+    * @param[in] Tr element transformation
+    * @param[in] elfun local coefficient of basis
+    * @param[out] elvect evaluated dual vector
+    */
+   void AssembleElementVector(const FiniteElement &el,
+                              ElementTransformation &Tr,
+                              const Vector &elfun, Vector &elvect) override;
+
+   /**
+    * @brief implement <-hat(F)(u,x) n, [[v]]> with abstract hat(F) computed by
+    * ComputeFluxDotN and numerical flux object
+    *
+    * @param[in] el1 finite element of the first element
+    * @param[in] el2 finite element of the second element
+    * @param[in] Tr face element transformations
+    * @param[in] elfun local coefficient of basis from both elements
+    * @param[out] elvect evaluated dual vector <-hat(F)(u,x) n, [[v]]>
+    */
+   void AssembleFaceVector(const FiniteElement &el1,
+                           const FiniteElement &el2,
+                           FaceElementTransformations &Tr,
+                           const Vector &elfun, Vector &elvect) override;
+
+};
+
+
+// Base Hyperbolic conservation law class.
+// This contains all methods needed except the flux function.
+class DGHyperbolicConservationLaws : public TimeDependentOperator
+{
+private:
+   const int dim;
+   const int num_equations;
+   // Vector finite element space containing conserved variables
+   FiniteElementSpace *vfes;
+   // Element integration form. Should contain ComputeFlux
+   HyperbolicFormIntegrator *formIntegrator;
+   // Base Nonlinear Form
+   std::unique_ptr<NonlinearForm> nonlinearForm;
+   // element-wise inverse mass matrix
+   std::vector<DenseMatrix> Me_inv;
+   // global maximum characteristic speed. Updated by form integrators
+   mutable double max_char_speed;
+   // auxiliary variable used in Mult
+   mutable Vector z;
+
+   // Compute element-wise inverse mass matrix
+   void ComputeInvMass();
+
+   void Update();
+
+public:
+   /**
+    * @brief Construct a new DGHyperbolicConservationLaws object
+    *
+    * @param vfes_ vector finite element space. Only tested for DG [Pₚ]ⁿ
+    * @param formIntegrator_ integrator (F(u,x), grad v)
+    */
+   DGHyperbolicConservationLaws(
+      FiniteElementSpace &vfes_,
+      HyperbolicFormIntegrator *formIntegrator_);
+   /**
+    * @brief Apply nonlinear form to obtain M⁻¹(DIVF + JUMP HAT(F))
+    *
+    * @param x current solution vector
+    * @param y resulting dual vector to be used in an EXPLICIT solver
+    */
+   void Mult(const Vector &x, Vector &y) const override;
+   // get global maximum characteristic speed to be used in CFL condition
+   // where max_char_speed is updated during Mult.
+   inline double getMaxCharSpeed()
```

Comment:
```suggestion
   double GetMaxCharSpeed()
```
Since the definition is inside the class body, this is automatically `inline`.